src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java

```diff
@@ -39,195 +50,289 @@
     @Override
     @CheckForNull
-    protected SCMRevision retrieve(@NonNull String thingName, @NonNull TaskListener listener) throws IOException, InterruptedException {
-        MercurialInstallation inst = MercurialSCM.findInstallation(installation);
-        if (inst == null) {
-            listener.error("No configured Mercurial installation");
-            return null;
-        }
-        if (!inst.isUseCaches()) {
-            listener.error("Mercurial installation " + installation + " does not support caches");
-            return null;
-        }
-        final Node node = Jenkins.getActiveInstance();
-        Launcher launcher = node.createLauncher(listener);
-        StandardUsernameCredentials credentials = getCredentials();
-        final FilePath cache = Cache.fromURL(source, credentials, inst.getMasterCacheRoot()).repositoryCache(inst, node, launcher, listener, true);
-        if (cache == null) {
-            listener.error("Could not use caches, not fetching branch heads");
-            return null;
-        }
-        final HgExe hg = new HgExe(inst, credentials, launcher, node, listener, new EnvVars());
-        try {
-            String revision = hg.popen(cache, listener, true, new ArgumentListBuilder("log", "-r", "present("+thingName+")", "--template", "{node} {branch}"));
-            if (revision.isEmpty()){
+    protected SCMRevision retrieve(@NonNull String thingName, @NonNull TaskListener listener)
+            throws IOException, InterruptedException {
+        try (MercurialSCMSourceRequest request = new MercurialSCMSourceContext<>(null, SCMHeadObserver.none())
+                .withTraits(traits)
+                .newRequest(this, listener)) {
+            MercurialInstallation inst = MercurialSCM.findInstallation(request.installation());
+            if (inst == null) {
+                listener.error("No configured Mercurial installation");
                 return null;
             }
-            String hash = revision.substring(0, revision.indexOf(' '));
-            String branch = revision.substring(revision.indexOf(' ')+1, revision.length());
-            return new MercurialRevision(new SCMHead(branch), hash);
-        } finally {
-            hg.close();
+            if (!inst.isUseCaches()) {
+                listener.error("Mercurial installation " + request.installation() + " does not support caches");
+                return null;
+            }
+            final Node node = Jenkins.getInstance();
+            if (node == null) { // Should not happen BTW
+                listener.error("Cannot retrieve the Jenkins master node");
+                return null;
+            }
+            Launcher launcher = node.createLauncher(listener);
+            StandardUsernameCredentials credentials = getCredentials(request.credentialsId());
+            final FilePath cache = Cache.fromURL(source, credentials, inst.getMasterCacheRoot())
+                    .repositoryCache(inst, node, launcher, listener, true);
```

Comment:
Did you somehow contrive to make this diff as long as possible? Would much rather see this diff hunk disappear, and the `try` block simply not be indented.